pkg/controller/jobframework/reconciler.go

```diff
@@ -1151,6 +1192,47 @@ func ConstructWorkload(ctx context.Context, c client.Client, job GenericJob, lab
 	return wl, nil
 }
 
+// prepareWorkloadSlice adds necessary workload slice annotations.
+func prepareWorkloadSlice(ctx context.Context, clnt client.Client, job GenericJob, wl *kueue.Workload) error {
+	// Lookup existing slice for a given job.
+	workloadSlices, err := workloadslicing.FindNotFinishedWorkloads(ctx, clnt, job.Object(), job.GVK())
+	if err != nil {
+		return fmt.Errorf("failure looking up workload slices: %w", err)
+	}
+
+	switch len(workloadSlices) {
+	case 0:
+		// No active workloads found for this job - noop.
+		return nil
+	case 1:
+		// One active workload found for this job - typically, we are in a scale-up event, where previous
+		// workload is the "old" slice.
+		oldSlice := workloadSlices[0]
+		// Assert that we are processing a new workload-slice.
+		if wl.Name == oldSlice.Name {
```

Comment:
As you say this should never happen, so I don't think it makes sense to keep the dead code around. I would suggest to drop it.